File: sdk/src/main/scala/com/horizen/account/abi/ABIListEncoder.scala
package com.horizen.account.abi

import org.web3j.abi.DefaultFunctionEncoder
import org.web3j.abi.datatypes.{DynamicArray, StaticStruct}
import org.web3j.utils.Numeric

import java.util
import scala.collection.JavaConverters.seqAsJavaListConverter

trait ABIListEncoder[M <: ABIEncodable] {
  val encoder = new DefaultFunctionEncoder()

  def encode(listOfObj: Seq[M]): Array[Byte] = {

    val listOfABIObj = listOfObj.map(wr => wr.asABIType()).asJava
    Numeric.hexStringToByteArray(encoder.encodeParameters(util.Arrays.asList(new DynamicArray[StaticStruct](classOf[StaticStruct], listOfABIObj))))

  }

}

File: sdk/src/main/java/io/horizen/account/abi/ABIListEncoder.java
package io.horizen.account.abi;

import org.web3j.abi.DefaultFunctionEncoder;
import org.web3j.abi.datatypes.DynamicArray;
import org.web3j.abi.datatypes.Type;
import org.web3j.utils.Numeric;

import java.util.List;
import java.util.stream.Collectors;

public interface ABIListEncoder<M extends ABIEncodable<T>, T extends Type> {

    Class<T> getAbiClass();

    default byte[] encode(List<M> listOfObj){
        DefaultFunctionEncoder encoder = new DefaultFunctionEncoder();
        List<T> listOfABIObj = listOfObj.stream().map(wr -> wr.asABIType()).collect(Collectors.toList());

        return Numeric.hexStringToByteArray(encoder.encodeParameters(List.of(new DynamicArray<T>(getAbiClass(), listOfABIObj))));

    }
}
